Reintroduce new generic parsing system

DenTelezhkin opened the pull request:
This PR is a restart of #707, and a first step towards better generics handling by Sourcery.

Changes are intentionally smaller in scope than in the original, and are almost backwards-compatible. In fact, the only truly breaking change I see is `Type.isGeneric` property that became computed (previously it was passed in initializer), but it's unlikely that this change will break someone's usage of the framework.

This PR follows glossary and terminology proposed in #707, so it's better to read full explanation provided in that PR.

Brief summary of changes that were implemented here:
* Introduce `Type.genericTypePlaceholders` and `Type.genericTypeParameters` properties, fill them on FileParser and Composer level
* Introduce `Method.genericTypeParameters` property
* Preserve `Type` knowledge wherever possible, even if types are unknown - in scope of objects, that are related to generics
* Implement type specialization(for example `Observable<Element>` can be a template class, and `Observable<Int>` can be a specialized version of that class, that contains both generic type placeholder `Element` and generic type parameter `Int`)

Even though this PR is only the first step, and does not finish generics parsing fundamentally, it was designed to be a good incremental improvement to not turn this PR into breaking monster it was previously.
 
/cc @ilyapuchka @truizlop

ilyapuchka commented on SourceryFramework/Sources/Parsing/Composer.swift:
```diff
@@ -156,44 +174,53 @@ public struct Composer {
+        
+        let type = unique[resolvedTypeName] ?? typeFromModule(resolvedTypeName, modules: modules)
+        if let knownPresumedType = presumedType, knownPresumedType.isConcreteGenericType {
+            let resolvedGenericTypeParameters = knownPresumedType.genericTypeParameters.map {
```
I'd maybe move this map to `specializeType` as it's part of specialisation process anyway

DenTelezhkin replied:
The reason I did not move this to `specializeType` is that we would need to pass `resolveTypeWithName` there, which actually makes this method to not only specialize, but resolve types too. In current form `specializeType` does one thing and one thing only - takes existing type, and creates a new one with specified generic type parameters, and does not search for underlying types.

ilyapuchka commented on SourceryFramework/Sources/Parsing/Composer.swift:
```diff
@@ -348,6 +375,16 @@ public struct Composer {
+        // This hack with keyed archivers serves only one goal - copy type to specialize it with generic parameters
+        // We don't want to change original type, it should stay unspecialized
+        let data = NSKeyedArchiver.archivedData(withRootObject: type)
+        let typeCopy = NSKeyedUnarchiver.unarchiveObject(with: data) as? Type
```
I guess this is done for brevity and not checking for all the possible types of `Type`, but this will probably come with a performance hit? Maybe would be better to introduce our custom Copyable protocol and generate code for all subclasses of `Type` that would use plain `init`.

DenTelezhkin replied:
I agree, but i think this would be a bigger change, that needs to be resolved in a separate pull request. I'm adding a TODO item to those lines to address your comment.

ilyapuchka replied:
The problem is that I guess this may affect performance of processing source code for everyone even if generic types are not used after all. I can't tell for sure though, but we could check on the performance tests that that we should have? If there is a negative effect but we want to fix it separately maybe it will make sense to introduce "experimental" feature flag and do not parse generics unless it is enabled? Also considering that (I guess) generics support is still not complete for some things.

DenTelezhkin replied:
I tried to run perfomance tests in `Sourcery+PerformanceSpec`, but they seem to be broken in a lot of ways, not sure how to properly fix and run them.

I tried to see perfomance difference on one of my fairly large projects, but got hit with some issues with indirect enums, that crash Sourcery on my project when used. Unfortunately, I rebased on master, so all features currently in master are in Sourcery build i tried.

Experimental flag seems a good idea to me, since there are some other things that could be done with generics, however I have no idea how to properly implement a feature flag for Sourcery. 

if you can try to run perfomance tests on Sourcery and/or on one of bigger project you have that use Sourcery, that would be great.

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -152,7 +152,15 @@ public final class FileParser {
 
-            type.isGeneric = isGeneric(source: source)
+            if let suffix = extract(.nameSuffixUpToBody, from: source) {
+                let declaration = extractGenericsDeclaration(source: suffix)
```
```suggestion
let genericDeclaration = extractGenericsDeclaration(source: suffix)
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -152,7 +152,15 @@ public final class FileParser {
-            type.isGeneric = isGeneric(source: source)
+            if let suffix = extract(.nameSuffixUpToBody, from: source) {
+                let declaration = extractGenericsDeclaration(source: suffix)
+                if !declaration.isEmpty {
```
```suggestion
                if !genericDeclaration.isEmpty {
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -273,6 +281,21 @@ public final class FileParser {
 extension FileParser {
+    
+    fileprivate func parseConcreteType(source: String) -> Type? {
+        guard !source.isEmpty else { return nil }
```
```suggestion
        guard !source.isEmpty else { return nil }

```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -273,6 +281,21 @@ public final class FileParser {
+    
+    fileprivate func parseConcreteType(source: String) -> Type? {
+        guard !source.isEmpty else { return nil }
+        let declaration = extractGenericsDeclaration(source: source)
```
```suggestion
        let genericDeclaration = extractGenericsDeclaration(source: source)
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -152,7 +152,15 @@ public final class FileParser {
+            if let suffix = extract(.nameSuffixUpToBody, from: source) {
+                let declaration = extractGenericsDeclaration(source: suffix)
+                if !declaration.isEmpty {
+                    // This type is generic, parse it's requirements
```
```suggestion
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -273,6 +281,21 @@ public final class FileParser {
+        guard !source.isEmpty else { return nil }
+        let declaration = extractGenericsDeclaration(source: source)
+        if declaration.isEmpty {
+            // This is not generic type.
```
```suggestion
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -273,6 +281,21 @@ public final class FileParser {
+            return Type(name: source)
+        }
+        let typeName = extractGenericTypeName(source: source)
+        guard declaration.contains("<") else {
```
```suggestion
        guard genericDeclaration.contains("<") else {
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -273,6 +281,21 @@ public final class FileParser {
+        let typeName = extractGenericTypeName(source: source)
+        guard declaration.contains("<") else {
+            return Type(name: typeName,
+                        genericTypeParameters: extractGenericTypeParameters(source: declaration))
```
```suggestion
                        genericTypeParameters: extractGenericTypeParameters(source: genericDeclaration))
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -318,6 +336,144 @@ extension FileParser {
+    fileprivate func parseGenericPlaceholders(source: [String: SourceKitRepresentable], definedIn: Any?) {
+        guard let type = definedIn as? Type else { return }
+        guard let substructure = source[SwiftDocKey.substructure.rawValue] as? [[String: SourceKitRepresentable]] else { return }
+        substructure.forEach { dict in
```
```suggestion

        substructure.forEach { dict in
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -318,6 +336,144 @@ extension FileParser {
+                var inheritedTypesParsed: [Type] = []
+                inheritedTypes.forEach { constraintString in
+                    let constraints = constraintString.components(separatedBy: "&")
+                    inheritedTypesParsed.append(contentsOf: constraints.map { recursivelyParseGenericPlaceholderDeclaration(source: $0) })
```
```suggestion
                    inheritedTypesParsed.append(contentsOf: constraints.map { 
                        recursivelyParseGenericPlaceholderDeclaration(source: $0) 
                    })
```

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -318,6 +336,144 @@ extension FileParser {
+            let constraints = String(source[source.index(after: constraintStart) ..< source.endIndex])
+                .components(separatedBy: "&")
+                .map { constraint in
+                    recursivelyParseGenericPlaceholderDeclaration(source: constraint)
```
do we need to do that for each type? I think it's only allowed to use concrete type as a first type of composition and everything else would be protocols, i.e. `ViewController<A> & Presentable & Dismissible`. Probably would be a bit faster if we only ever do it on first element.

DenTelezhkin replied:
Unfortunately, we do. Unit test in FileParserSpec "extracts generics correctly" on line 270 shows specific example where this is needed.

ilyapuchka commented on SourceryFramework/Sources/Parsing/FileParser.swift:
```diff
@@ -318,6 +336,144 @@ extension FileParser {
+    }
+    
+    fileprivate func findGeneric(withName name: String, in array: [GenericTypePlaceholder]) -> GenericTypePlaceholder? {
+        let filtered = array.filter { generic in generic.placeholderName.name == name }
```
`filter` and `first` can now be replaced with `first(where:)`

ilyapuchka commented on SourceryTests/Parsing/ComposerSpec.swift:
```diff
@@ -396,6 +403,32 @@ class ParserComposerSpec: QuickSpec {
+                        """)
+                        let generic = Class(name: "Generic", genericTypePlaceholders: [ GenericTypePlaceholder(placeholderName: TypeName("T")) ])
+                        let specialized = Class(name: "Specialized", inheritedTypes: ["Generic<Int>"])
+                        expect(result).to(equal([ generic, specialized]))
```
```suggestion

                        expect(result).to(equal([generic, specialized]))
```

ilyapuchka commented on SourceryTests/Parsing/ComposerSpec.swift:
```diff
@@ -396,6 +403,32 @@ class ParserComposerSpec: QuickSpec {
+                            class Specialized : Generic<Int> {}
+                        """)
+                        let generic = Class(name: "Generic", genericTypePlaceholders: [ GenericTypePlaceholder(placeholderName: TypeName("T")) ])
+                        let specialized = Class(name: "Specialized", inheritedTypes: ["Generic<Int>"])
```
Can we also validate that `inherits` property is populated as expected and contains `Generic` type? Or this is not the case?

DenTelezhkin replied:
Currently, this property is not populated. Reason is in `Composer.updateTypeRelationships(types:)` method `typesByName` contains "Generic" and "Specialized", however `type.inheritedTypes` contains "Generic<Int>", and therefore we cannot find "Generic", since we are searching for "Generic<Int>". 

I have a feeling it's a part of a bigger problem with generic names, but not sure how to resolve this one in scope of current pull request. Would love your advice on this.

ilyapuchka commented on SourceryTests/Parsing/ComposerSpec.swift:
```diff
@@ -396,6 +403,32 @@ class ParserComposerSpec: QuickSpec {
+                                                       genericTypeParameters: [GenericTypeParameter(typeName: TypeName("Int"), type: Type(name: "Int"))])
+                        let placeholder = GenericTypePlaceholder(placeholderName: TypeName("U"), constraints: [specializedGeneric])
+                        let specialized = Class(name: "Specialized", genericTypePlaceholders: [placeholder])
+                        expect(result).to(equal([ generic, specialized]))
```
```suggestion
                        expect(result).to(equal([generic, specialized]))
```

ilyapuchka commented on SourceryTests/Parsing/ComposerSpec.swift:
```diff
@@ -396,6 +403,32 @@ class ParserComposerSpec: QuickSpec {
+
+                    it("resolves inherited types for generic placeholders") {
+                        let result = parse("""
+                            class Generic<T> {}
```
can we add a test where there is a constraint here as well, i.e. `T: Bar` where `protocol Bar {}`

ilyapuchka commented on SourceryTests/Parsing/FileParser+SubscriptsSpec.swift:
```diff
@@ -39,7 +39,7 @@ class FileParserSubscriptsSpec: QuickSpec {
                         Subscript(
                             parameters: [
-                                MethodParameter(argumentLabel: "b", name: "b", typeName: TypeName("Int"))
+                                MethodParameter(argumentLabel: "b", name: "b", typeName: TypeName("Int"), type: Type(name: "Int"))
```
now when we have standard library types inferred does it still work properly when we have an extension of this type. It used to be the only case when we would set `type`, can you make sure we have a test for that and it still works?

DenTelezhkin replied:
Sorry, I don't seem to fully understand what test do we need here. Can you expand on this a little bit?

ilyapuchka replied:
I probably meant something like what if we have an extension of Int in the scanned code? Will this type reference contain all symbols from this extension? Meaning do we set this type property to the fully resolved type?

DenTelezhkin replied:
Got it, expanded unit test a little bit to have an extension on Int, and it's set for subscripts in more places then before(even though Int is reported as extension, because it's not seen anywhere else).

DenTelezhkin commented:
Sorry for dragging this since forever. Was burned out on my daily job and couldn't force myself to finish this. If anyone wishes to continue this PR further in more reasonable time, please do that.

In the meantime, I rebased on master, and addressed code review comments, some of them fixed, some require additional discussion or separate PR.

DenTelezhkin commented:
Hi all!

Clearly, this PR did not get enough interest(even from me), so those improvements do not seem like anything worth pursuing. If somebody wants to continue working with those changes, you are welcome to do so, I will not delete fork in the foreseeable future, but I will not pursue this further. Thanks to everyone who reviewed and was interested in this PR.
